compile error match in if expression

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version: 0.2.2 d012f27** 
**OS:ArchLinux** 

**What did you do?**
```v
fn main() {
	num := 3
	str := 'a'
	res := if num in [1, 3] {
		match str {
			'a' { 'A' }
			else { 'ODD' }
		}
	} else {
		'NONE'
	}
	assert res == 'A'
}
```

**What did you expect to see?**
works
 
**What did you see instead?**
```
tmp/v/main.7957456820858073030.tmp.c: In function 'main__main':
/tmp/v/main.7957456820858073030.tmp.c:8444:50: error: expected ')' before '_t69'
 8444 |  string res = ((num == 1 || num == 3) ? (  string _t69 = str;
      |                                         ~        ^~~~~
      |                                                  )
/tmp/v/main.7957456820858073030.tmp.c:8445:68: error: expected expression before ':' token
 8445 |   ((string_eq(_t69, _SLIT("a"))) ? (_SLIT("A")) : (_SLIT("ODD")))) : (_SLIT("NONE")));
      |                                                                    ^
...
```

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -3444,21 +3444,30 @@ fn (mut g Gen) match_expr(node ast.MatchExpr) {
 		return
 	}
 	is_expr := (node.is_expr && node.return_type != table.void_type) || g.inside_ternary > 0
+	mut cond_var := ''
 	if is_expr {
 		g.inside_ternary++
 		// g.write('/* EM ret type=${g.typ(node.return_type)}		expected_type=${g.typ(node.expected_type)}  */')
 	}
-	cur_line := if is_expr {
-		g.empty_line = true
-		g.go_before_stmt(0)
+	if node.cond is ast.Ident || node.cond is ast.SelectExpr || node.cond is ast.IndexExpr {
+		pos := g.out.len
+		g.expr(node.cond)
+		cond_var = g.out.after(pos)
+		g.out.go_back(cond_var.len)
 	} else {
-		''
+		cur_line := if is_expr {
+			g.empty_line = true
+			g.go_before_stmt(0)
+		} else {
+			''
+		}
+		cond_var = g.new_tmp_var()
+		g.write('${g.typ(node.cond_type)} $cond_var = ')
+		g.expr(node.cond)
+		g.writeln('; ')
+		g.write(cur_line)
 	}
-	cond_var := g.new_tmp_var()
-	g.write('${g.typ(node.cond_type)} $cond_var = ')
-	g.expr(node.cond)
-	g.writeln('; ')
-	g.write(cur_line)
+
 	if is_expr {
 		// brackets needed otherwise '?' will apply to everything on the left
 		g.write('(')

--- a/vlib/v/tests/match_in_if_expression_test.v
+++ b/vlib/v/tests/match_in_if_expression_test.v
@@ -1,0 +1,13 @@
+fn test_match_in_if_expression() {
+	num := 3
+	str := 'a'
+	res := if num in [1, 3] {
+		match str {
+			'a' { 'A' }
+			else { 'ODD' }
+		}
+	} else {
+		'NONE'
+	}
+	assert res == 'A'
+}

